File: back-end/src/main/java/es/ull/project/domain/valueobject/capacity/StorageCapacityKilograms.java
package es.ull.project.domain.valueobject.capacity;

import java.util.Objects;

/**
 * StorageCapacityKilograms
 *
 * Represents the maximum storage capacity of a facility in kilograms.
 * Immutable value object that encapsulates the storage capacity value in kilograms.
 * The unit is always kilograms (fixed).
 */
public final class StorageCapacityKilograms {

    /**
     * Error messages for validation.
     */
    private static final String ERROR_CAPACITY_NEGATIVE = "Storage capacity cannot be negative";
    private static final String ERROR_OTHER_CAPACITY_NULL = "Other StorageCapacityKilograms cannot be null";
    private static final int ZERO = 0;

    /**
     * Required.
     * Storage capacity value in kilograms.
     * It is a required attribute and always measured in kilograms.
     */
    private final double kilograms;

    /**
     * Creates a new StorageCapacityKilograms.
     *
     * @param kilograms Storage capacity value in kilograms (must be ≥ 0)
     * @throws IllegalArgumentException if kilograms is negative
     */
    public StorageCapacityKilograms(double kilograms) {
        if (kilograms < ZERO) {
            throw new IllegalArgumentException(ERROR_CAPACITY_NEGATIVE);
        }
        this.kilograms = kilograms;
    }

    /**
     * Returns the storage capacity value in kilograms.
     *
     * @return Storage capacity value in kilograms
     */
    public double getKilograms() {
        return this.kilograms;
    }

    /**
     * Returns a new StorageCapacityKilograms with the updated value.
     *
     * @param newKilograms New storage capacity value in kilograms
     * @return New StorageCapacityKilograms
     * @throws IllegalArgumentException if newKilograms is negative
     */
    public StorageCapacityKilograms setKilograms(double newKilograms) {
        return new StorageCapacityKilograms(newKilograms);
    }

    /**
     * Compares if this capacity is greater than another.
     *
     * @param other Other StorageCapacityKilograms
     * @return true if greater, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean greaterThan(StorageCapacityKilograms other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_CAPACITY_NULL);
        }
        return this.kilograms > other.kilograms;
    }

    /**
     * Compares if this capacity is greater than or equal to another.
     *
     * @param other Other StorageCapacityKilograms
     * @return true if greater or equal, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean greaterThanOrEqual(StorageCapacityKilograms other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_CAPACITY_NULL);
        }
        return this.kilograms >= other.kilograms;
    }

    /**
     * Compares if this capacity is less than another.
     *
     * @param other Other StorageCapacityKilograms
     * @return true if less, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean lessThan(StorageCapacityKilograms other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_CAPACITY_NULL);
        }
        return this.kilograms < other.kilograms;
    }

    /**
     * Compares equality of two StorageCapacityKilograms instances.
     *
     * @param otherObject Other object to compare
     * @return true if equal, false otherwise
     */
    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (otherObject == null || getClass() != otherObject.getClass()) {
            return false;
        }
        StorageCapacityKilograms other = (StorageCapacityKilograms) otherObject;
        return Double.compare(this.kilograms, other.kilograms) == ZERO;
    }

    /**
     * Returns the hash code of the StorageCapacityKilograms.
     *
     * @return hashCode
     */
    @Override
    public int hashCode() {
        return Objects.hash(kilograms);
    }

    /**
     * Returns the string representation of the StorageCapacityKilograms.
     *
     * @return String
     */
    @Override
    public String toString() {
        return String.format("StorageCapacityKilograms={kilograms=%s}", this.kilograms);
    }
}

File: front-end/src/domain/valueobject/capacity/storage-capacity-kilograms.ts
/**
 * StorageCapacityKilograms
 *
 * Represents the maximum storage capacity of a facility in kilograms.
 * Immutable value object that encapsulates the capacity value in kilograms.
 * The unit is always kilograms (fixed).
 */
export class StorageCapacityKilograms {
  /**
   * Capacity value in kilograms.
   * Must be non-negative.
   */
  private readonly kilograms: number;

  /**
   * Create a new `StorageCapacityKilograms`.
   * @param kilograms The capacity in kilograms (must be >= 0)
   * @throws Error if kilograms is negative
   */
  constructor(kilograms: number) {
    if (kilograms < 0) {
      throw new Error('Storage capacity cannot be negative');
    }
    this.kilograms = kilograms;
  }

  /**
   * Get the capacity value in kilograms.
   * @returns The capacity in kilograms
   */
  getKilograms(): number {
    return this.kilograms;
  }

  /**
   * Create a new instance with an updated value.
   * @param newKilograms The new capacity value in kilograms
   * @returns A new StorageCapacityKilograms instance
   * @throws Error if newKilograms is negative
   */
  setKilograms(newKilograms: number): StorageCapacityKilograms {
    return new StorageCapacityKilograms(newKilograms);
  }

  /**
   * Compare if this capacity is greater than another.
   * @param other The other StorageCapacityKilograms to compare
   * @returns true if this capacity is greater, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThan(other: StorageCapacityKilograms): boolean {
    if (!other) {
      throw new Error('Other StorageCapacityKilograms cannot be null');
    }
    return this.kilograms > other.kilograms;
  }

  /**
   * Compare if this capacity is greater than or equal to another.
   * @param other The other StorageCapacityKilograms to compare
   * @returns true if this capacity is greater or equal, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThanOrEqual(other: StorageCapacityKilograms): boolean {
    if (!other) {
      throw new Error('Other StorageCapacityKilograms cannot be null');
    }
    return this.kilograms >= other.kilograms;
  }

  /**
   * Compare if this capacity is less than another.
   * @param other The other StorageCapacityKilograms to compare
   * @returns true if this capacity is less, false otherwise
   * @throws Error if other is null or undefined
   */
  lessThan(other: StorageCapacityKilograms): boolean {
    if (!other) {
      throw new Error('Other StorageCapacityKilograms cannot be null');
    }
    return this.kilograms < other.kilograms;
  }

  /**
   * Compare equality with another object.
   * @param otherObject The object to compare with
   * @returns true if the objects are equal, false otherwise
   */
  equals(otherObject: unknown): boolean {
    if (this === otherObject) {
      return true;
    }
    if (!(otherObject instanceof StorageCapacityKilograms)) {
      return false;
    }
    return this.kilograms === otherObject.kilograms;
  }

  /**
   * Get the hash code for this object.
   * @returns A hash code based on the kilograms value
   */
  hashCode(): number {
    return this.kilograms !== 0 ? this.kilograms : 0;
  }

  /**
   * Get the string representation of this object.
   * @returns A string representation
   */
  toString(): string {
    return `StorageCapacityKilograms={kilograms=${this.kilograms}}`;
  }
}
